lib/iris/experimental/ugrid/mesh.py

```diff
@@ -1962,6 +1962,292 @@ def topology_dimension(self):
+        face_requested = include_faces
+
+        # Rationalise the tri-state behaviour.
+        args = [include_nodes, include_edges, include_faces]
+        state = not any(set(filter(lambda arg: arg is not None, args)))
+        include_nodes, include_edges, include_faces = map(
+            lambda arg: arg if arg is not None else state, args
+        )
+
+        def populated_coords(coords_tuple):
+            return list(filter(None, list(coords_tuple)))
+
+        members = []
+        if include_nodes:
+            members += populated_coords(self.node_coords)
+        if include_edges:
+            members += populated_coords(self.edge_coords)
+        if hasattr(self, "face_coords"):
+            if include_faces:
+                members += populated_coords(self.face_coords)
+        elif face_requested:
+            dmsg = "Ignoring request to filter non-existent 'face_coords'"
+            logger.debug(dmsg, extra=dict(cls=self.__class__.__name__))
+
+        result = metadata_filter(
+            members,
+            item=item,
+            standard_name=standard_name,
+            long_name=long_name,
+            var_name=var_name,
+            attributes=attributes,
+            axis=axis,
+        )
+
+        # Use the results to filter the _members dict for returning.
+        result_ids = [id(r) for r in result]
+        result_dict = {k: v for k, v in self._members.items() if id(v) in result_ids}
+        return result_dict
+
+    def filter(self, **kwargs):
+        # TODO: rationalise commonality with MeshConnectivityManager.filter and Cube.coord.
+        result = self.filters(**kwargs)
+
+        if len(result) > 1:
+            names = ", ".join(f"{member}={coord!r}" for member, coord in result.items())
+            emsg = (
+                f"Expected to find exactly 1 coordinate, but found {len(result)}. "
+                f"They were: {names}."
+            )
+            raise CoordinateNotFoundError(emsg)
+
+        if len(result) == 0:
+            item = kwargs["item"]
+            if item is not None:
+                if not isinstance(item, str):
+                    item = item.name()
+            name = (
+                item
+                or kwargs["standard_name"]
+                or kwargs["long_name"]
+                or kwargs["var_name"]
+                or None
+            )
+            name = "" if name is None else f"{name!r} "
+            emsg = f"Expected to find exactly 1 {name}coordinate, but found none."
+            raise CoordinateNotFoundError(emsg)
+
+        return result
+
+
+class _MeshIndexConnectivityManager(_MeshIndexManager):
+    @property
+    def edge_node(self):
+        if self.edge_indices is None:
+            return None
+        else:
+            connectivity = self.mesh.edge_node_connectivity[self.edge_indices]
+            connectivity_indices = np.vectorize(self.node_index_dict.get)(
+                connectivity.indices
+            )
```

Comment:
So I think the problem with storing this as an array where the one set of indices is infered by the position is that the indices aren't guaranteed to be of the form `[0,1,2,3,...]` but may have gaps in them. It ought to be possible to rewrite this to be of the form `connectivity_indices = index_array[connectivity.indices]`, but this would require building `index_array` to be about as large as the coordinates on the original mesh rather than scaling with the size of the provided indices. It's probably still worth checking how well this performs, because this may well be the more natural way to use numpy.